### PR TITLE
Track holes to be able to reuse them once LFC limit is increased

### PR DESCRIPTION
**Pull request by knizhnik**
## Problem

Multiple increase/decrease LFC limit may cause unlimited growth of LFC file because punched holes while LFC shrinking are not reused when LFC is extended.

## Summary of changes

Keep track of holes and reused them when LFC size is increased.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by kelvich** [comment]
I'd suggest to change chunk size in a separate release to reduce number of dangerous changes in a single release.

As for the patch itself -- there is a bunch of connected constants scattered through the code, e.g. 32 and (i >> 5) can't be changed independently. Can we reuse postgres bitmapset's https://github.com/postgres/postgres/blob/master/src/include/nodes/bitmapset.h ?

**Comment by hlinnaka** [comment]
My first thought was: I find it surprising to use a bitmap to track the free chunks. We already allocate a FileCacheEntry for each chunk, whether it's in use or not. You can have a freelist for free, since oyu

**Comment by hlinnaka** [comment]
Oops, clicked "Close with comment" by accident... Sorry, reopened

**Comment by hlinnaka** [comment]
My first thought was: I find it surprising to use a bitmap to track the free chunks. We already allocate a FileCacheEntry for each chunk, whether it's in use or not. You can have a freelist for free, because you already have a FreeCacheEntry with a 'dlist_node' available for each chunk.

Then I realized that it's not quite that simple: The FileCacheEntry's are allocated and freed by dynahash, as we enter and remove entries from the hash table. So we don't really have a FileCacheEntry available for chunks that are not in use. Dynahash keeps a freelist of entries, but that's not accessible to us.

An alternative would be to allocate all the FileCacheEntry's ourselves, and use simplehash instead of dynahash. There are some pros and cons: it's generally faster, but won't allow us to partition the hash table so easily, if we want to do so in the future.

One unstated point of the bitmap is that it prefers to reuse holes from the beginning of the file. Not sure how important that is.

Another alternative design would be to always truncate the file at the end, throwing away whatever chunks were stored there. If there were some hot chunks, they will need to be read from the pageserver on next access, but it's probably acceptable performance-wise.

**Comment by hlinnaka** [comment]
> Then I realized that it's not quite that simple: The FileCacheEntry's are allocated and freed by dynahash, as we enter and remove entries from the hash table. So we don't really have a FileCacheEntry available for chunks that are not in use. Dynahash keeps a freelist of entries, but that's not accessible to us.

Hmm, one way is to add each hole to the hash table, with a dummy key that doesn't clash with any real buffer tags. Here's a prototype with that approach: https://github.com/neondatabase/neon/compare/main...lfc-holes. It passes the lfc_resize.py test from this PR, but haven't done any testing otherwise.

**Comment by knizhnik** [comment]
> Hmm, one way is to add each hole to the hash table, with a dummy key that doesn't clash with any real buffer tags. Here's a prototype with that approach: https://github.com/neondatabase/neon/compare/main...lfc-holes. It passes the lfc_resize.py test from this PR, but haven't done any testing otherwise.

Yes, certainly it was my first idea just to maintain yet another list of holes. But then I also realised that entries are deallocated. by dynhash. Certainly it is possible to propose own hash implementation which doesn't deallocate entries.
Or as you suggested use dummy key. But it seems to me that bitmap is the simplest and most straightforward solution.
Extra memory overhead (but per chunk) is negligible. Other drawbacks? CP overhead for scanning bitmap? Using `last_hole` pointer it is also should not be noticeable. And bitmaps provides one really useful feature: access locality. Holes are filled in the order of their location in file. Which should slightly increase access locality and so performance in most cases.

So I still think that implementation with bitmap is better alternative.

**Comment by hlinnaka** [comment]
What was the reason for changing the chunk size? (I agree with Stas that that should be a separate PR, whatever the reason was)

**Comment by hlinnaka** [comment]
> > Hmm, one way is to add each hole to the hash table, with a dummy key that doesn't clash with any real buffer tags. Here's a prototype with that approach: [main...lfc-holes](https://github.com/neondatabase/neon/compare/main...lfc-holes). It passes the lfc_resize.py test from this PR, but haven't done any testing otherwise.
> 
> Yes, certainly it was my first idea just to maintain yet another list of holes. But then I also realised that entries are deallocated. by dynhash. Certainly it is possible to propose own hash implementation which doesn't deallocate entries. Or as you suggested use dummy key. But it seems to me that bitmap is the simplest and most straightforward solution. Extra memory overhead (but per chunk) is negligible. Other drawbacks? CPU overhead for scanning bitmap? Using `last_hole` pointer it is also should not be noticeable. And bitmaps provides one really useful feature: access locality. Holes are filled in the order of their location in file. Which should slightly increase access locality and so performance in most cases.
> 
> So I still think that implementation with bitmap is better alternative.

Feels about a wash to me. The implementation in this PR is pretty hard to follow, and the list approach therefore looks easier, but I think the bitmap approach could be made much more readable too, with some helper functions, constants and macros to abstract the bitmap manipulation.

**Comment by knizhnik** [comment]
> What was the reason for changing the chunk size? (I agree with Stas that that should be a separate PR, whatever the reason was)

See https://neondb.slack.com/archives/C04DGM6SMTM/p1722399136350679
@sharnoff noticed very strange effect: after increasing LFC size, size of OS buffer is decreased.
Looks lie I know the answer. It is pgbench with almost random access pattern and working set much larger than size of LFC. It means that each 1Mb chunk contains just one of few pages.  When page eviction is performed, we do not punch hole (yet another option to be consider). So even through pages of this chunk are not used any more, they are still kept in OS cache. But when we change LFC size, then. we use empty chunks and now OS cache contains only used pages.

I assume that most of customers have OLTP-like workload. So random access pattern isomer likely than sequential access pattern for OLAP queries. It is not clear to me whether current LFC size (depends on CU but <= 30Gb) is enough to fir working set for most of customers. But if working set doesn't and we have trashing then small chunk will  be much more efficient. As I already said, at pgbench with scale 1000 and 30Gb LFC cache, I got about 2x time improvement of performance with 64kb chunk.

Big chunks was chosen for LFC because of three reasons:
1. To minimise memory footprint when size of LFC is very larger (terabytes).
2. Optimize sequential scan (subsequent  pages are also located together in the disk)
3. Minimize fragmentation caused by punching holes.

With current LFC limit (100Gb), size of LFC with 64kb chunks will be ~hundred of megabytes. With 1Mb pages it is 16x times smaller - few megabytes.

Ok, I will restore previous chunk value in this PR.

**Comment by knizhnik** [comment]
> Feels about a wash to me. The implementation in this PR is pretty hard to follow, and the list approach therefore looks easier, but I think the bitmap approach could be made much more readable too, with some helper functions, constants and macros to abstract the bitmap manipulation.

Do you mean that extra locality provided by bitmap can not somehow affect performance?
I am not sure that it will be visible, but usually file systems and DBMSes are specially using allocation bitmaps for this purpose.

Concerning making bitmap manipulation more readable: I will do if you agree with this approach rather than using hole tag keys. Otherwise let's merge your proposal..

**Comment by sharnoff** [comment]
> With current LFC limit (100Gb) ...

FYI with neondatabase/cloud#15940 the LFC limit is now equal to 80% of max CU.

**Comment by hlinnaka** [comment]
> > Feels about a wash to me. The implementation in this PR is pretty hard to follow, and the list approach therefore looks easier, but I think the bitmap approach could be made much more readable too, with some helper functions, constants and macros to abstract the bitmap manipulation.
> 
> Do you mean that extra locality provided by bitmap can not somehow affect performance? I am not sure that it will be visible, but usually file systems and DBMSes are specially using allocation bitmaps for this purpose.

I doubt it makes any measurable difference in this case. The file can be pretty badly fragmented with holes all over the file anyway, because we don't try to defragment in any way except when it grows.

> Concerning making bitmap manipulation more readable: I will do if you agree with this approach rather than using hole tag keys. Otherwise let's merge your proposal..

I'd recommend proceeding with the hole tag keys, but I don't feel strongly about it. As I said, I think it's about a wash, both can be made to work.

**Comment by knizhnik** [comment]
Ok, I used your approach with hole tag keys.
The minor change: I replaced LIFO with FIFO.

**Comment by knizhnik** [comment]
Some results of local LFC performance testing:
```
shared_buffers=128MB
file_cacher_size_limit=1GB
pgbench scale=600
pgbench -c 10 -T 100 -S -M prepared
```

| chunk size | punch on evict |  TPC |    hits   |  misses  | ls -lh   |
|------------|----------------|------|-------|----------|-------|
|.   1Mb         |.          no            | 7039 |     0    | 1368891 |  1.1Gb |
|.   1Mb         |.         yes            | 3005 |     0    | 585453 |  11Mb  |
|.  64kb.       |            no            | 8209 |16531| 1579604|   1.1Gb |
|.  64kb        |          yes            | 3273 |   6391| 630938 | 160Mb |

Looks like punching hole on chunk eviction is not good idea - it reduces speed almost twice.

File: pgxn/neon/file_cache.c [hunk 1]
```diff
@@ -51,28 +51,52 @@
...
+ * fallocate(FALLOC_FL_PUNCH_HOLE) call. The nominal size of the file doesn't
+ * shrink, but the disk space it uses does.
+ *
+ * Each hole is tracked by a dummy FileCacheEntry, which are kept in the
+ * 'holes' linked list. They are entered into the chunk hash table, with a
+ * special key where the blockNumber is used to store the 'offset' of the
+ * hole, and all other fields are zero. Holes are never looked up in the hash
+ * table, we only enter them there to have a FileCacheEntry that we can keep
+ * in the linked list. If the soft limit is raised again, we reuse the holes
+ * before extending the nominal size of file.
```
Comment (hlinnaka) [inline pgxn/neon/file_cache.c:72]:
```suggestion
 * before extending the nominal size of the file.
```

File: pgxn/neon/file_cache.c [hunk 2]
```diff
@@ -51,28 +51,52 @@
...
  * 1. Reduce hash-map memory footprint: 8TB database contains billion pages
  *    and size of hash entry is 40 bytes, so we need 40Gb just for hash map.
  *    1Mb chunks can reduce hash map size to 320Mb.
  * 2. Improve access locality, subsequent pages will be allocated together improving seqscan speed
  */
 #define BLOCKS_PER_CHUNK	128 /* 1Mb chunk */
+/*
+ * Smaller chunk seems to be better for OLTP workload
+ */
+// #define BLOCKS_PER_CHUNK	8 /* 64kb chunk */
```
Comment (hlinnaka) [inline pgxn/neon/file_cache.c:86]: What's this?

File: pgxn/neon/file_cache.c [hunk 3]
```diff
@@ -85,8 +109,9 @@ typedef struct FileCacheControl
 	uint64		hits;
 	uint64		misses;
 	uint64		writes;
-	dlist_head	lru;			/* double linked list for LRU replacement
+ 	dlist_head	lru;			/* double linked list for LRU replacement
```
Comment (hlinnaka) [inline pgxn/neon/file_cache.c:112]: stray space

File: pgxn/neon/file_cache.c [hunk 4]
```diff
@@ -124,6 +149,7 @@ lfc_disable(char const *op)
 	{
 		HASH_SEQ_STATUS status;
 		FileCacheEntry *entry;
+		size_t n_chunks = SIZE_MB_TO_CHUNKS(lfc_max_size);
```
Comment (hlinnaka) [inline pgxn/neon/file_cache.c:152]: this is unused

File: pgxn/neon/file_cache.c [hunk 5]
```diff
@@ -655,13 +707,26 @@ lfc_write(NRelFileInfo rinfo, ForkNumber forkNum, BlockNumber blkno, const void
...
+			FileCacheEntry *victim = dlist_container(FileCacheEntry, list_node, dlist_pop_head_node(&lfc_ctl->lru));
 
 			Assert(victim->access_count == 0);
 			entry->offset = victim->offset; /* grab victim's chunk */
 			hash_search_with_hash_value(lfc_hash, &victim->key, victim->hash, HASH_REMOVE, NULL);
 			neon_log(DEBUG2, "Swap file cache page");
 		}
+		else if (!dlist_is_empty(&lfc_ctl->holes))
+		{
+			/* We can reuse a hole that was left behind when the LFC was shrink previously */
```
Comment (hlinnaka) [inline pgxn/neon/file_cache.c:719]:
```suggestion
			/* We can reuse a hole that was left behind when the LFC was shrunk previously */
```

**Review by hlinnaka — APPROVED** [review]
Hmm, I think we should change most of the Asserts to elog(PANIC) checks, so that we check all the invariants even in release builds. Because if the LFC data structures ever get corrupt for some reason, that's likely to cause very hard-to-debug data corruption. Better to detect that early and stop the server.

File: pgxn/neon/file_cache.c (hunk 1 above)
Comment (knizhnik) [reply pgxn/neon/file_cache.c:72]: Thank you , fixed.

File: pgxn/neon/file_cache.c (hunk 3 above)
Comment (knizhnik) [reply pgxn/neon/file_cache.c:112]: Fixed

File: pgxn/neon/file_cache.c (hunk 4 above)
Comment (knizhnik) [reply pgxn/neon/file_cache.c:152]: Removed

File: pgxn/neon/file_cache.c (hunk 5 above)
Comment (knizhnik) [reply pgxn/neon/file_cache.c:719]: Corrected

**Review by ololobus — APPROVED** [review]
I had a test with a lot of scales up and down running over ~24h -- https://neondb.slack.com/archives/C03H1K0PGKH/p1723754163642499?thread_ts=1723648356.641969&cid=C03H1K0PGKH

Didn't see any issues. Also did amcheck today on this DB

We also should've finished a few prodlike benchmark runs this week, so I think it's good to go
